threadpoolctl.py

```diff
@@ -60,56 +67,322 @@ class _dl_phdr_info(ctypes.Structure):
     _RTLD_NOLOAD = ctypes.DEFAULT_MODE
 
 
-# List of the supported libraries. The items are indexed by the name of the
-# class to instantiate to create the library controller objects. The items hold
-# the possible prefixes of loaded shared objects, the name of the internal_api
-# to call, the name of the user_api and potentially some symbols that the library is
-# expected to have (this is necessary to distinguish between the blas implementations
-# when they are all renamed "libblas.dll" on conda-forge on windows).
-_SUPPORTED_LIBRARIES = {
-    "OpenMPController": {
-        "user_api": "openmp",
-        "internal_api": "openmp",
-        "filename_prefixes": ("libiomp", "libgomp", "libomp", "vcomp"),
-    },
-    "OpenBLASController": {
-        "user_api": "blas",
-        "internal_api": "openblas",
-        "filename_prefixes": ("libopenblas", "libblas"),
-        "check_symbols": ("openblas_get_num_threads", "openblas_get_num_threads64_"),
-    },
-    "MKLController": {
-        "user_api": "blas",
-        "internal_api": "mkl",
-        "filename_prefixes": ("libmkl_rt", "mkl_rt", "libblas"),
-        "check_symbols": ("MKL_Get_Max_Threads",),
-    },
-    "BLISController": {
-        "user_api": "blas",
-        "internal_api": "blis",
-        "filename_prefixes": ("libblis", "libblas"),
-        "check_symbols": ("bli_thread_get_num_threads",),
-    },
-}
+class LibController(ABC):
+    """Abstract base class for the individual library controllers
+
+    A library controller must expose the following class attributes:
+        - user_api : str
+            Usually the name of the library or generic specification the library
+            implements.
```

Comment:
Maybe explain why it is used to give hints on how to select them?
```suggestion
            implements. This allows limiting generic frameworks
            regardless of the specific implementations, e.g. openMP.
```

I am actually not sure this one should be mandatory, maybe it could default to `internal_api`, or the other way around?

Reply:
It's doable but it will add some code complexity because we do check both attrs sometimes before the class is instanciated. I think it's ok to keep both mandatory, one can repeat the same name anyway.
We can still decide to make it optional later, but I'd give this a try first if it's fine with you.